tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredFileSegment.java

```diff
@@ -0,0 +1,529 @@
+            tmpBufferList.addAll(uploadBufferList);
+            uploadBufferList = tmpBufferList;
+            if (inputStream.getCodaBuffer() != null) {
+                codaBuffer.rewind();
+            }
+        } finally {
+            bufferLock.unlock();
+        }
+    }
+
+    public AppendResult append(ByteBuffer byteBuf, long timeStamp) {
+        if (closed) {
+            return AppendResult.FILE_CLOSE;
+        }
+        bufferLock.lock();
+        try {
+            if (full || codaBuffer != null) {
+                return AppendResult.FILE_FULL;
+            }
+
+            if (fileType == FileSegmentType.INDEX) {
+                beginTimestamp = byteBuf.getLong(TieredIndexFile.INDEX_FILE_HEADER_BEGIN_TIME_STAMP_POSITION);
+                endTimestamp = byteBuf.getLong(TieredIndexFile.INDEX_FILE_HEADER_END_TIME_STAMP_POSITION);
+                appendPosition += byteBuf.remaining();
+                uploadBufferList.add(byteBuf);
+                setFull();
+                return AppendResult.SUCCESS;
+            }
+
+            if (appendPosition + byteBuf.remaining() > maxSize) {
+                setFull();
+                return AppendResult.FILE_FULL;
+            }
+            if (uploadBufferList.size() > storeConfig.getTieredStoreGroupCommitCount()
+                || appendPosition - commitPosition > storeConfig.getTieredStoreGroupCommitSize()) {
+                commitAsync();
+            }
+            if (uploadBufferList.size() > storeConfig.getTieredStoreMaxGroupCommitCount()) {
+                LOGGER.debug("TieredFileSegment#append: buffer full: file: {}, upload buffer size: {}",
+                    getPath(), uploadBufferList.size());
+                return AppendResult.BUFFER_FULL;
+            }
+            if (timeStamp != Long.MAX_VALUE) {
+                endTimestamp = timeStamp;
+                if (beginTimestamp == Long.MAX_VALUE) {
+                    beginTimestamp = timeStamp;
+                }
+            }
+            appendPosition += byteBuf.remaining();
+            uploadBufferList.add(byteBuf);
+            return AppendResult.SUCCESS;
+        } finally {
+            bufferLock.unlock();
+        }
+    }
+
+    private void appendCoda() {
+        if (codaBuffer != null) {
+            return;
+        }
+        codaBuffer = ByteBuffer.allocate(TieredCommitLog.CODA_SIZE);
+        codaBuffer.putInt(TieredCommitLog.CODA_SIZE);
+        codaBuffer.putInt(TieredCommitLog.BLANK_MAGIC_CODE);
+        codaBuffer.putLong(endTimestamp);
+        codaBuffer.flip();
+        appendPosition += TieredCommitLog.CODA_SIZE;
+    }
+
+    public ByteBuffer read(long position, int length) {
+        return readAsync(position, length).join();
+    }
+
+    public CompletableFuture<ByteBuffer> readAsync(long position, int length) {
+        if (position < 0 || length < 0) {
+            throw new TieredStoreException(TieredStoreErrorCode.ILLEGAL_PARAM, "position or length is negative");
```

Comment:
The `future` here implies that there may be an exception, could we return `future` directly?